lib/Service/TokenService.php

```diff
@@ -0,0 +1,276 @@
+<?php
+
+declare(strict_types=1);
+/**
+ * SPDX-FileCopyrightText: 2024 Nextcloud GmbH and Nextcloud contributors
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+namespace OCA\UserOIDC\Service;
+
+use OCA\UserOIDC\AppInfo\Application;
+use OCA\UserOIDC\Db\ProviderMapper;
+use OCA\UserOIDC\Exception\TokenExchangeFailedException;
+use OCA\UserOIDC\Model\Token;
+use OCA\UserOIDC\Vendor\Firebase\JWT\JWT;
+use OCP\AppFramework\Db\DoesNotExistException;
+use OCP\AppFramework\Db\MultipleObjectsReturnedException;
+use OCP\Http\Client\IClient;
+use OCP\Http\Client\IClientService;
+use OCP\IConfig;
+use OCP\IRequest;
+use OCP\ISession;
+use OCP\IURLGenerator;
+use OCP\IUserSession;
+use OCP\PreConditionNotMetException;
+use OCP\Security\ICrypto;
+use Psr\Log\LoggerInterface;
+
+/**
+ * Token management service
+ * This is helpful to debug:
+ * tail -f data/nextcloud.log | grep "\[Token" | jq ".time,.message"
+ */
+class TokenService {
+
+	private const SESSION_TOKEN_KEY = Application::APP_ID . '-user-token';
+
+	private IClient $client;
+
+	public function __construct(
+		IClientService $clientService,
+		private ISession $session,
+		private IUserSession $userSession,
+		private IConfig $config,
+		private LoggerInterface $logger,
+		private ICrypto $crypto,
+		private IRequest $request,
+		private IURLGenerator $urlGenerator,
+		private DiscoveryService $discoveryService,
+		private ProviderMapper $providerMapper,
+	) {
+		$this->client = $clientService->newClient();
+	}
+
+	public function storeToken(array $tokenData): Token {
+		$token = new Token($tokenData);
+		$this->session->set(self::SESSION_TOKEN_KEY, json_encode($token, JSON_THROW_ON_ERROR));
+		$this->logger->debug('[TokenService] Store token');
+		return $token;
+	}
+
+	/**
+	 * Get the token stored in the session
+	 * If it has expired: try to refresh it
+	 *
+	 * @param bool $refreshIfExpired
+	 * @return Token|null Return a token only if it is valid or has been successfully refreshed
+	 * @throws \JsonException
+	 */
+	public function getToken(bool $refreshIfExpired = true): ?Token {
+		$sessionData = $this->session->get(self::SESSION_TOKEN_KEY);
+		if (!$sessionData) {
+			$this->logger->debug('[TokenService] getToken: no session data');
```

Comment:
Should this be logged at a higher level?

Reply:
Maybe it will flood the logs since this one is produced not only when there's a token exchange request but also when checking the login token in Application.php. If the current user is not authenticated via user_oidc, there will never be any session data.

Reply:
All right then debug is also fine.